src/java.base/share/classes/java/util/concurrent/Joiners.java

```diff
@@ -0,0 +1,220 @@
+            }
+        }
+    }
+
+    /**
+     * A joiner that returns the result of the first subtask to complete successfully.
+     * Cancels the scope if any subtasks succeeds.
+     */
+    static final class AnySuccessful<T> implements Joiner<T, T> {
+        private static final VarHandle SUBTASK =
+                MhUtil.findVarHandle(MethodHandles.lookup(), "subtask", Subtask.class);
+
+        // UNAVAILABLE < FAILED < SUCCESS
+        private static final Comparator<Subtask.State> SUBTASK_STATE_COMPARATOR =
+                Comparator.comparingInt(AnySuccessful::stateToInt);
+
+        private volatile Subtask<T> subtask;
+
+        /**
+         * Maps a Subtask.State to an int that can be compared.
+         */
+        private static int stateToInt(Subtask.State s) {
+            return switch (s) {
+                case UNAVAILABLE -> 0;
+                case FAILED      -> 1;
+                case SUCCESS     -> 2;
+            };
+        }
+
+        @Override
+        public boolean onComplete(Subtask<? extends T> subtask) {
+            Subtask.State state = ensureCompleted(subtask);
+            Subtask<T> s;
+            while (((s = this.subtask) == null)
+                    || SUBTASK_STATE_COMPARATOR.compare(s.state(), state) < 0) {
+                if (SUBTASK.compareAndSet(this, s, subtask)) {
+                    return (state == Subtask.State.SUCCESS);
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public T result() throws Throwable {
+            Subtask<T> subtask = this.subtask;
+            if (subtask == null) {
+                throw new NoSuchElementException("No subtasks completed");
+            }
+            return switch (subtask.state()) {
+                case SUCCESS -> subtask.get();
+                case FAILED -> throw subtask.exception();
+                default -> throw new InternalError();
+            };
+        }
+    }
+
+    /**
+     * A joiner that that waits for all successful subtasks. Cancels the scope if any
+     * subtask fails.
+     */
+    static final class AwaitSuccessful<T> implements Joiner<T, Void> {
+        private static final VarHandle FIRST_EXCEPTION =
+                MhUtil.findVarHandle(MethodHandles.lookup(), "firstException", Throwable.class);
+        private volatile Throwable firstException;
+
+        @Override
+        public boolean onComplete(Subtask<? extends T> subtask) {
+            Subtask.State state = ensureCompleted(subtask);
+            return (state == Subtask.State.FAILED)
+                    && (firstException == null)
+                    && FIRST_EXCEPTION.compareAndSet(this, null, subtask.exception());
+        }
+
+        @Override
+        public Void result() throws Throwable {
+            Throwable ex = firstException;
+            if (ex != null) {
+                throw ex;
+            } else {
+                return null;
+            }
```

Comment:
```suggestion
            if (ex != null) {
                throw ex;
            }
            return null;
```
The else here is redundant